File: backend/Bot/OrderExecutor.cs
using Microsoft.EntityFrameworkCore;
using TradingBot.Banks;
using TradingBot.Data;
using TradingBot.Indicators;
using TradingBot.Models;

namespace TradingBot.Bot;

public class OrderExecutor(AppDbContext db, BankAdapterFactory bankFactory, IConfiguration config, ILogger<OrderExecutor> logger)
{
    private static bool IsWithinTradingHours()
    {
        try
        {
            // Windows: "Turkey Standard Time", Linux: "Europe/Istanbul"
            var tzId = OperatingSystem.IsWindows() ? "Turkey Standard Time" : "Europe/Istanbul";
            var tz = TimeZoneInfo.FindSystemTimeZoneById(tzId);
            var turkeyTime = TimeZoneInfo.ConvertTimeFromUtc(DateTime.UtcNow, tz);
            return turkeyTime.Hour >= 10 && turkeyTime.Hour < 18;
        }
        catch
        {
            // Fallback: UTC+3
            var turkeyTime = DateTime.UtcNow.AddHours(3);
            return turkeyTime.Hour >= 10 && turkeyTime.Hour < 18;
        }
    }

    public async Task ExecuteAsync(Guid userId, string ticker, AggregatedSignal signal, double maxOrderSizeTry, double dailyLossLimitTry)
    {
        bool paperTrade = config.GetValue<bool>("Trading:PaperTradingMode", true);

        if (!IsWithinTradingHours())
        {
            logger.LogDebug("Outside BIST trading hours, skipping {Ticker}", ticker);
            return;
        }

        // Daily loss check (simplified)
        var today = DateTime.UtcNow.Date;
        double dailyLoss = await db.Orders
            .Where(o => o.UserId == userId && o.CreatedAt >= today && o.Status == OrderStatus.FILLED && o.Type == OrderType.SELL)
            .SumAsync(o => o.Price * o.Quantity * 0.002);

        if (dailyLoss >= dailyLossLimitTry)
        {
            logger.LogWarning("Daily loss limit reached for user {UserId}", userId);
            return;
        }

        var symbol = await db.BistSymbols.FirstOrDefaultAsync(s => s.Ticker == ticker);
        if (symbol == null || symbol.LastPrice <= 0) return;

        var orderType = signal.Signal == SignalType.BUY ? OrderType.BUY : OrderType.SELL;
        int quantity = (int)(maxOrderSizeTry / symbol.LastPrice);
        if (quantity <= 0) return;

        if (orderType == OrderType.SELL)
        {
            var position = await db.Positions.FirstOrDefaultAsync(p => p.UserId == userId && p.SymbolId == symbol.Id);
            if (position == null || position.Quantity <= 0) return;
        }

        var adapter = bankFactory.GetDefault();
        var bankOrder = new BankOrder(ticker, orderType, quantity, symbol.LastPrice, paperTrade);

        var order = new Order
        {
            UserId = userId,
            SymbolId = symbol.Id,
            Type = orderType,
            Quantity = quantity,
            Price = symbol.LastPrice,
            Status = OrderStatus.PENDING,
            BankAdapter = adapter.Name,
            PaperTrade = paperTrade,
            Notes = $"Auto: confidence={signal.Confidence:F3}",
        };
        db.Orders.Add(order);
        await db.SaveChangesAsync();

        try
        {
            var result = await adapter.PlaceOrderAsync(bankOrder);
            order.Status = result.Status;
            order.Price = result.FilledPrice;
            order.FilledAt = result.FilledAt;
            await db.SaveChangesAsync();

            if (result.Status == OrderStatus.FILLED)
                await UpsertPositionAsync(userId, symbol.Id, orderType, quantity, result.FilledPrice);

            logger.LogInformation("{Type} {Qty} {Ticker} @ {Price:F2} — {Status}", orderType, quantity, ticker, result.FilledPrice, result.Status);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Order execution failed for {Ticker}", ticker);
            order.Status = OrderStatus.FAILED;
            await db.SaveChangesAsync();
        }
    }

    private async Task UpsertPositionAsync(Guid userId, Guid symbolId, OrderType type, double qty, double price)
    {
        var pos = await db.Positions.FirstOrDefaultAsync(p => p.UserId == userId && p.SymbolId == symbolId);

        if (type == OrderType.BUY)
        {
            if (pos != null)
            {
                double totalQty = pos.Quantity + qty;
                pos.AvgPrice = (pos.Quantity * pos.AvgPrice + qty * price) / totalQty;
                pos.Quantity = totalQty;
                pos.UpdatedAt = DateTime.UtcNow;
            }
            else
            {
                db.Positions.Add(new Position { UserId = userId, SymbolId = symbolId, Quantity = qty, AvgPrice = price });
            }
        }
        else if (pos != null)
        {
            pos.Quantity -= qty;
            pos.UpdatedAt = DateTime.UtcNow;
            if (pos.Quantity <= 0) db.Positions.Remove(pos);
        }

        await db.SaveChangesAsync();
    }
}

File: backend/src/bot/orderExecutor.ts
import { PrismaClient } from '@prisma/client';
import { AggregatedSignal, BankOrder, OrderType } from '../types';
import { getDefaultAdapter } from '../banks/factory';

const prisma = new PrismaClient();

function isWithinTradingHours(): boolean {
  const now = new Date();
  // Turkey is UTC+3
  const utcMs = now.getTime() + now.getTimezoneOffset() * 60_000;
  const turkeyTime = new Date(utcMs + 3 * 3_600_000);
  const hour = turkeyTime.getHours();
  const minute = turkeyTime.getMinutes();
  const totalMinutes = hour * 60 + minute;
  // 10:00 to 18:00 Turkey time
  return totalMinutes >= 600 && totalMinutes < 1080;
}

export async function executeOrder(
  userId: string,
  ticker: string,
  signal: AggregatedSignal,
  maxOrderSizeTry: number,
  dailyLossLimitTry: number
): Promise<void> {
  const paperTrade = process.env.PAPER_TRADING_MODE !== 'false';

  if (!isWithinTradingHours()) {
    console.log(`[OrderExecutor] Outside BIST trading hours, skipping ${ticker}`);
    return;
  }

  // Check daily loss
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const todayOrders = await prisma.order.findMany({
    where: {
      userId,
      createdAt: { gte: today },
      status: 'FILLED',
      type: 'SELL',
    },
  });

  // Simplified daily loss check (based on filled sell orders)
  const dailyLoss = todayOrders.reduce((sum, o) => sum + o.price * o.quantity * 0.002, 0);
  if (dailyLoss >= dailyLossLimitTry) {
    console.log(`[OrderExecutor] Daily loss limit reached for user ${userId}`);
    return;
  }

  const symbol = await prisma.bistSymbol.findUnique({ where: { ticker } });
  if (!symbol) {
    console.error(`[OrderExecutor] Symbol not found: ${ticker}`);
    return;
  }

  const price = symbol.lastPrice;
  if (price <= 0) return;

  const orderType: OrderType = signal.signal === 'BUY' ? 'BUY' : 'SELL';
  const quantity = Math.floor(maxOrderSizeTry / price);
  if (quantity <= 0) return;

  // For SELL, check if user has position
  if (orderType === 'SELL') {
    const position = await prisma.position.findUnique({
      where: { userId_symbolId: { userId, symbolId: symbol.id } },
    });
    if (!position || position.quantity <= 0) {
      console.log(`[OrderExecutor] No position to sell for ${ticker}`);
      return;
    }
  }

  const adapter = getDefaultAdapter();
  const bankOrder: BankOrder = {
    symbolTicker: ticker,
    type: orderType,
    quantity,
    price,
    paperTrade,
  };

  const order = await prisma.order.create({
    data: {
      userId,
      symbolId: symbol.id,
      type: orderType,
      quantity,
      price,
      status: 'PENDING',
      bankAdapter: adapter.name,
      paperTrade,
      notes: `Auto: confidence=${signal.confidence.toFixed(3)}`,
    },
  });

  try {
    const result = await adapter.placeOrder(bankOrder);

    await prisma.order.update({
      where: { id: order.id },
      data: {
        status: result.status,
        price: result.filledPrice,
        filledAt: result.filledAt,
      },
    });

    if (result.status === 'FILLED') {
      if (orderType === 'BUY') {
        await upsertPosition(userId, symbol.id, quantity, result.filledPrice);
      } else {
        await reducePosition(userId, symbol.id, quantity);
      }
    }

    console.log(`[OrderExecutor] ${orderType} ${quantity} ${ticker} @ ${result.filledPrice.toFixed(2)} — ${result.status}`);
  } catch (err) {
    console.error(`[OrderExecutor] Failed to execute order for ${ticker}:`, err);
    await prisma.order.update({ where: { id: order.id }, data: { status: 'FAILED' } });
  }
}

async function upsertPosition(userId: string, symbolId: string, qty: number, price: number): Promise<void> {
  const existing = await prisma.position.findUnique({
    where: { userId_symbolId: { userId, symbolId } },
  });

  if (existing) {
    const totalQty = existing.quantity + qty;
    const newAvg = (existing.quantity * existing.avgPrice + qty * price) / totalQty;
    await prisma.position.update({
      where: { userId_symbolId: { userId, symbolId } },
      data: { quantity: totalQty, avgPrice: newAvg },
    });
  } else {
    await prisma.position.create({ data: { userId, symbolId, quantity: qty, avgPrice: price } });
  }
}

async function reducePosition(userId: string, symbolId: string, qty: number): Promise<void> {
  const existing = await prisma.position.findUnique({
    where: { userId_symbolId: { userId, symbolId } },
  });

  if (!existing) return;
  const remaining = existing.quantity - qty;
  if (remaining <= 0) {
    await prisma.position.delete({ where: { userId_symbolId: { userId, symbolId } } });
  } else {
    await prisma.position.update({
      where: { userId_symbolId: { userId, symbolId } },
      data: { quantity: remaining },
    });
  }
}
